src/main/java/edu/kit/compiler/optimizations/CallGraph.java

```diff
@@ -0,0 +1,234 @@
+        return Visitor.create(graphs);
+    }
+
+    /**
+     * Update the given function in the call graph. All calls to the function
+     * will remain in the call graph, all outgoing calls will be reevaluated.
+     * The function is not required to currently exist in the call graph (may be
+     * the case if an optimization adds new functions to the program).
+     * 
+     * Note: This operation is only valid for functions with associated graph,
+     * hence the parameter is a `Graph` instead of an `Entity`.
+     */
+    public void update(firm.Graph function) {
+        this.components = null;
+        Visitor.update(this, function);
+    }
+
+    private Components getOrInitComponents() {
+        if (components != null) {
+            return components;
+        } else {
+            return (components = new Components(graph));
+        }
+    }
+
+    @Override
+    public String toString() {
+        return String.format("CallGraph(functions=%s, calls=%s", graph.vertexSet(),
+                graph.edgeSet().stream().map(this::formatEdge)
+                        .collect(Collectors.toSet()));
+    }
+
+    private String formatEdge(DefaultEdge edge) {
+        return String.format("%s -> %s",
+                graph.getEdgeSource(edge),
+                graph.getEdgeTarget(edge));
+    }
+
+    private static Entity getCaller(Call call) {
+        return call.getGraph().getEntity();
+    }
+
+    private static Entity getCallee(Call call) {
+        if (call.getPtr().getOpCode() == ir_opcode.iro_Address) {
+            return ((Address) call.getPtr()).getEntity();
+        } else {
+            throw new IllegalStateException("only constant function pointers allowed");
+        }
+    }
+
+    private static final class Components {
+
+        private final List<Set<Entity>> components;
+        private final Map<Entity, Set<Entity>> functionMap;
+
+        public Components(Graph<Entity, DefaultEdge> graph) {
+            var sccInspector = new KosarajuStrongConnectivityInspector<>(graph);
+            var connectedSets = sccInspector.stronglyConnectedSets();
+
+            this.components = connectedSets;
+            this.functionMap = new HashMap<>();
+
+            for (var component : components) {
+                for (var node : component) {
+                    functionMap.put(node, component);
+                }
+            }
+        }
+
+        public boolean isSameComponent(Entity func1, Entity func2) {
+            var comp1 = functionMap.get(func1);
+            var comp2 = functionMap.get(func2);
+
+            // compare components using equality operator
+            return comp1 != null && comp1 == comp2;
+        }
+    }
+
+    @RequiredArgsConstructor
+    @AllArgsConstructor
+    private static final class Visitor extends NodeVisitor.Default {
+
+        private Entity caller;
+        private final Graph<Entity, DefaultEdge> graph;
+
+        public static CallGraph create() {
```

Comment:
Are two `create` functions necessary here? You could simply call `Visitor.create(Program.getGraphs())` in `CallGraph::create` and have a cleaner interface for `Visitor`.